Warn about floor_divide performing incorrect rounding

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50438 Support JIT shape analysis for div with rounding_mode argument
* #51989 Sparse support for division rounding_mode argument
* #51792 Vectorize floating point floor_divide
* #51988 Replace rounding_mode="true" with rounding_mode=None
* **#50281 Warn about floor_divide performing incorrect rounding**

Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

peterbell10 committed:
Warn about floor_divide performing incorrect rounding

[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -319,6 +319,13 @@ Tensor& remainder_(Tensor& self, const Tensor& other) {
 
 Tensor& floor_divide_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  TORCH_WARN_ONCE(
+    "floor_divide rounding will change in a future version of pytorch. "
```
"floor_divide is deprecated, and it will be removed in a future version of PyTorch. It currently..."

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -319,6 +319,13 @@ Tensor& remainder_(Tensor& self, const Tensor& other) {
+    "floor_divide currently rounds toward 0 (like the 'trunc' function NOT 'floor'). "
+    "This results in incorrect rounding for negative values.\n"
+    "To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), "
+    "or for true floor division, use torch.div(a, b, rounding_mode='floor')."
```
"true" -> "actual" to avoid confusion with "true" division

mruberry commented on aten/src/ATen/native/cuda/Indexing.cu:
```diff
@@ -203,7 +203,7 @@ void index_put_accum_kernel(Tensor & self, const c10::List<c10::optional<Tensor>
       const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
 
-      linearIndex.floor_divide_(sliceSize);
+      linearIndex.divide_(sliceSize, "trunc");
```
Are these the only internal floor_divide uses? 

Removing it and trying to compile PyTorch should reveal the remaining .cpp uses.

The Python uses can be detected by throwing a runtime error when the function is called. We should ensure there are no stray uses of a deprecated function, not even in the test suite.

peterbell10 replied:
>Removing it and trying to compile PyTorch should reveal the remaining .cpp uses.

Removing from `native_functions.yaml` revealed no ther uses. I used `git grep` to find these initially, so should be fairly complete.

>The Python uses can be detected by throwing a runtime error when the function is called.

I found a couple more call sites using this method.

mruberry replied:
Feel free to submit to the Github CI with the function throwing a runtime error so it will run the entire test suite for you to identify floor_divide uses in Python.

peterbell10 replied:
Submitted to CI and the only non-obvious calls are arising because of the JIT tracer turning `Tensor.size()` into a list of scalar tensors. These should probably be warning though, since they have been silently giving different answers when run through JIT.

mruberry commented on test/test_torch.py:
```diff
@@ -6548,9 +6548,11 @@ def inner(self, device, dtype):
         lambda t, d: [_small_3d(t, d, has_zeros=False)], 1e-1,
         1e-1, 1e-5, torch.testing.get_all_fp_dtypes()),
-    ('floor_divide', '', _small_3d, lambda t, d: [_number(3.14, 3, t)], 1, 1e-5, 1e-5, _types),
+    ('floor_divide', '', _small_3d, lambda t, d: [_number(3.14, 3, t)], 1, 1e-5, 1e-5, _types, (), True,
```
Let's implement a floor_divide OpInfo and remove these.

peterbell10 replied:
Done. Had to implement a new `supports_autograd` attribute to `OpInfo` since `floor_divide` isn't currently differentiable.

mruberry replied:
That sounds like good metadata.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3117,7 +3117,7 @@ def merge_dicts(*dicts):
     This function's name is a misnomer. It actually rounds the
     quotient towards zero instead of taking its floor. This behavior
-    will be deprecated in a future PyTorch release.
+    is deprecated and will change in a future PyTorch release.
```
Let's start this paragraph with the most important bit: 

"torch.floor_divide is deprecated and will be removed in a future PyTorch release. Its name is a misnomer because it actually rounds the quotient towards zero instead of taking its floor. To keep the current behavior use torch.div() with the 'trunc' rounding mode. To actually perform floor division use torch.div() with the 'floor' rounding mode."

mruberry commented on torch/csrc/jit/frontend/builtin_functions.cpp:
```diff
@@ -97,15 +97,15 @@ auto div_tensor = R"SCRIPT(
   if (self.is_floating_point() or other.is_floating_point()):
     return self.true_divide(other)
-  return self.floor_divide(other)
+  return self.divide(other, rounding_mode='trunc')
```
Thank you for updating these, too.

mruberry reviewed (commented):
Hey @peterbell10,

Another thorough PR. I made a few comments inline. Big points:

- let's OpInfo floor_divide
- docs update to reflect the function is deprecated and will be removed
- ensure we're updating all internal call sites to stop using this function

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"




[ghstack-poisoned]

mruberry reviewed (approved):
Looks good -- this will need a few rebase tweaks. More on that later.

mruberry commented:
Internal updates reflected in: https://github.com/pytorch/pytorch/pull/51745#issuecomment-773620116

mruberry commented:
Hey @peterbell10, take a look at the CI on https://github.com/pytorch/pytorch/pull/51745, which is this PR rebased and with a docs fix (deprecated -> warning).

It seems like there's some new issue with the ONNX testing where it's being passed a div signature it doesn't understand. I'm going to investigate more now.

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL. Rebased and implemented `rounding_mode` argument in the ONNX export. That fixes the ONNX test failures.

mruberry commented:
fyi: the pytorch_doc_test failure was in the base (and has since been addressed). It's not an issue with this PR.

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -320,6 +320,13 @@ Tensor& remainder_(Tensor& self, const Tensor& other) {
 }
 
 Tensor& floor_divide_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  TORCH_WARN_ONCE(
```
Take a look at https://github.com/pytorch/pytorch/pull/51745/files. This warning got a slight update and needs to be replicated for floor_divide(). You'll have to verify manually that the warning appears for the function, method, inplace, and out= variants.

mruberry commented on torch/onnx/symbolic_opset10.py:
```diff
@@ -20,6 +20,44 @@
 # release on 04/24/19
 
 
+def div(g, self, other, *args):
```
The ONNX changes need a test, too. See here:

https://github.com/pytorch/pytorch/blob/fcf8b712348f21634044a5d76a69a59727756357/test/onnx/test_pytorch_onnx_onnxruntime.py#L1308

The test does not need to be extensive, it just needs to validate that the call works correctly.

mruberry reviewed (commented):
Thanks for the updates, @peterbell10; I'll start testing this internally. I made two notes:

- check the updated warning message requirements and validate manually that the warning message is thrown as expected (see inline comment)
- add tests sanity checking the ONNX behavior (see inline comment for a code reference)

peterbell10 committed:
Update on "Warn about floor_divide performing incorrect rounding"


Differential Revision: [D26257855](https://our.internmc.facebook.com/intern/diff/D26257855)

[ghstack-poisoned]

peterbell10 commented:
@mruberry added onnx tests, ensured warning was included on both overloads and rebased to fix the doc issue (although there seems to be a new ROCm issue now).
